File: sdks/python/roost/resources/manifests.py
"""``roost.manifests`` — manifest detail + file listing + diff."""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from roost.client import RoostClient


class Manifests:
    def __init__(self, client: "RoostClient") -> None:
        self._client = client

    async def list(
        self,
        roost_id: str,
        *,
        site_id: str,
        page_size: int = 20,
    ) -> AsyncIterator[dict[str, Any]]:
        cursor: str | None = None
        while True:
            resp = await self._client.request(
                f"/api/roosts/{roost_id}/manifests",
                query={"siteId": site_id, "limit": page_size, "cursor": cursor},
            )
            data = resp.data if isinstance(resp.data, dict) else {}
            for m in data.get("manifests", []):
                if isinstance(m, dict):
                    yield m
            cursor = data.get("nextCursor")
            if not cursor:
                return

    async def get(
        self, roost_id: str, manifest_id: str, *, site_id: str
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/roosts/{roost_id}/manifests/{manifest_id}",
            query={"siteId": site_id},
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def files(
        self,
        roost_id: str,
        manifest_id: str,
        *,
        site_id: str,
        page_size: int = 100,
    ) -> AsyncIterator[dict[str, Any]]:
        cursor: str | None = None
        while True:
            resp = await self._client.request(
                f"/api/roosts/{roost_id}/manifests/{manifest_id}/files",
                query={"siteId": site_id, "limit": page_size, "cursor": cursor},
            )
            data = resp.data if isinstance(resp.data, dict) else {}
            for f in data.get("files", []):
                if isinstance(f, dict):
                    yield f
            next_token = str(data.get("nextPageToken") or "")
            if not next_token:
                return
            cursor = next_token

    async def diff(
        self,
        roost_id: str,
        manifest_id: str,
        *,
        site_id: str,
        against: str,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/roosts/{roost_id}/manifests/{manifest_id}/diff",
            query={"siteId": site_id, "against": against},
        )
        return resp.data if isinstance(resp.data, dict) else {}


__all__ = ["Manifests"]

File: sdks/node/src/resources/manifests.ts
import type { RoostClient } from '../lib/client';

export interface ManifestDetail {
  manifestId: string;
  roostId: string;
  siteId: string;
  manifest: Record<string, unknown>;
  metadata: {
    manifestUrl: string | null;
    createdAt: string | null;
    createdBy: string | null;
    totalSize: number;
    totalFiles: number;
    parentManifestId: string | null;
  };
}

export interface ManifestFilesPage {
  manifestId: string;
  roostId: string;
  siteId: string;
  total: number;
  files: Array<{ path: string; size: number; chunks: Array<{ hash: string; size: number }> }>;
  nextPageToken: string;
}

export interface ManifestDiff {
  manifestId: string;
  against: string;
  roostId: string;
  siteId: string;
  summary: {
    added: number;
    removed: number;
    changed: number;
    unchanged: number;
    hasChanges: boolean;
    netBytesDelta: number;
  };
  added: Array<{ path: string; size: number; reason: 'added'; chunks: number }>;
  removed: Array<{ path: string; size: number; reason: 'removed'; chunks: number }>;
  modified: Array<{
    path: string;
    fromSize: number;
    toSize: number;
    reason: 'modified';
    fromChunks: number;
    toChunks: number;
  }>;
}

export class Manifests {
  constructor(private readonly client: RoostClient) {}

  async list(
    roostId: string,
    opts: { siteId: string; limit?: number; cursor?: string },
  ): Promise<{ manifests: Array<Record<string, unknown>>; nextCursor: string | null }> {
    const res = await this.client.request<{
      manifests: Array<Record<string, unknown>>;
      nextCursor: string | null;
    }>(`/api/roosts/${encodeURIComponent(roostId)}/manifests`, {
      query: { siteId: opts.siteId, limit: opts.limit, cursor: opts.cursor },
    });
    return res.data;
  }

  async get(
    roostId: string,
    manifestId: string,
    opts: { siteId: string },
  ): Promise<ManifestDetail> {
    const res = await this.client.request<ManifestDetail>(
      `/api/roosts/${encodeURIComponent(roostId)}/manifests/${encodeURIComponent(manifestId)}`,
      { query: { siteId: opts.siteId } },
    );
    return res.data;
  }

  async files(
    roostId: string,
    manifestId: string,
    opts: { siteId: string; limit?: number; cursor?: string },
  ): Promise<ManifestFilesPage> {
    const res = await this.client.request<ManifestFilesPage>(
      `/api/roosts/${encodeURIComponent(roostId)}/manifests/${encodeURIComponent(manifestId)}/files`,
      {
        query: { siteId: opts.siteId, limit: opts.limit, cursor: opts.cursor },
      },
    );
    return res.data;
  }

  async diff(
    roostId: string,
    manifestId: string,
    opts: { siteId: string; against: string },
  ): Promise<ManifestDiff> {
    const res = await this.client.request<ManifestDiff>(
      `/api/roosts/${encodeURIComponent(roostId)}/manifests/${encodeURIComponent(manifestId)}/diff`,
      { query: { siteId: opts.siteId, against: opts.against } },
    );
    return res.data;
  }
}
